drivers/display/display_gc9x01x.c

```diff
@@ -0,0 +1,745 @@
+
+		if (retval) {
+			LOG_ERR("Failed to set pulse width\n");
+			return retval;
+		}
+		return 0;
+	}
+
+	return 0;
+}
+
+static int gc9x01x_init(const struct device *dev)
+{
+	const struct gc9x01x_config *config = dev->config;
+
+	int retval;
+
+	if (!spi_is_ready_dt(&config->spi)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	if (!gpio_is_ready_dt(&config->cmd_data)) {
+		LOG_ERR("Command/Data GPIO device not ready");
+		return -ENODEV;
+	}
+
+	retval = gpio_pin_configure_dt(&config->cmd_data, GPIO_OUTPUT);
+	if (retval < 0) {
+		LOG_ERR("Could not configure command/data GPIO (%d)", retval);
+		return retval;
+	}
+
+	if (config->reset.port != NULL) {
+		if (!device_is_ready(config->reset.port)) {
+			LOG_ERR("Reset GPIO device not ready");
+			return -ENODEV;
+		}
+
+		retval = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_INACTIVE);
+		if (retval < 0) {
+			LOG_ERR("Could not configure reset GPIO (%d)", retval);
+			return retval;
+		}
+	}
+
+	gc9x01x_hw_reset(dev);
+
+	gc9x01x_display_blanking_on(dev);
+
+	retval = gc9x01x_configure(dev);
+	if (retval < 0) {
+		LOG_ERR("Could not configure display (%d)", retval);
+		return retval;
+	}
+
+	retval = gc9x01x_exit_sleep(dev);
+	if (retval < 0) {
+		LOG_ERR("Could not exit sleep mode (%d)", retval);
+		return retval;
+	}
+
+	retval = gc9x01x_set_brightness(dev, 50);
+	if (retval < 0) {
+		LOG_ERR("Could not set brightness (%d)", retval);
+		return retval;
+	}
+
+	return 0;
+}
+
+static int gc9x01x_set_mem_area(const struct device *dev, const uint16_t x, const uint16_t y,
+				const uint16_t w, const uint16_t h)
+{
+	int retval;
+	uint16_t spi_data[2];
+
+	spi_data[0] = sys_cpu_to_be16(x);
+	spi_data[1] = sys_cpu_to_be16(x + w - 1U);
+	retval = gc9x01x_transmit(dev, GC9X01X_CMD_COLSET, &spi_data[0], 4U);
+	if (retval < 0) {
+		return retval;
+	}
+
+	spi_data[0] = sys_cpu_to_be16(y);
+	spi_data[1] = sys_cpu_to_be16(y + h - 1U);
+	retval = gc9x01x_transmit(dev, GC9X01X_CMD_ROWSET, &spi_data[0], 4U);
+	if (retval < 0) {
+		return retval;
+	}
+
+	return 0;
+}
+
+static int gc9x01x_write(const struct device *dev, const uint16_t x, const uint16_t y,
+			 const struct display_buffer_descriptor *desc, const void *buf)
+{
+	const struct gc9x01x_config *config = dev->config;
+	struct gc9x01x_data *data = dev->data;
+
+	int retval;
+	const uint8_t *write_data_start = (const uint8_t *)buf;
+	struct spi_buf tx_buf;
+	struct spi_buf_set tx_bufs;
+	uint16_t write_cnt;
+	uint16_t nbr_of_writes;
+	uint16_t write_h;
+
+	__ASSERT(desc->width <= desc->pitch, "Pitch is smaller than width");
+	__ASSERT((desc->pitch * data->bytes_per_pixel * desc->height) <= desc->buf_size,
+		 "Input buffer to small");
+
+	LOG_DBG("Writing %dx%d (w,h) @ %dx%d (x,y)", desc->width, desc->height, x, y);
+	retval = gc9x01x_set_mem_area(dev, x, y, desc->width, desc->height);
+	if (retval < 0) {
+		return retval;
+	}
+
+	if (desc->pitch > desc->width) {
+		write_h = 1U;
+		nbr_of_writes = desc->height;
```

Comment:
I'm wondering when the pitch can be greater than width? And why are we switching in line by line writes in this case?

Reply:
From my understanding pitch can be used to skip areas in the provided buffer (e.g. for interlacing or just copying a subpart of a loaded resource. In this case pitch can be larger than width. The remaining bytes in a pixel column therefore is calculated based on pitch instead of width.